Applications/Spire/Source/Ui/EditableTableView.cpp

```diff
@@ -0,0 +1,431 @@
+  }
+
+  const TableHeaderItem::Model& get(int index) const override {
+    if(index == 0 || index == get_size() - 1) {
+      static auto model = TableHeaderItem::Model{"", "",
+        TableHeaderItem::Order::UNORDERED, TableFilter::Filter::NONE};
+      return model;
+    }
+    return m_source->get(index - 1);
+  }
+
+  QValidator::State set(int index,
+      const TableHeaderItem::Model& value) override {
+    if(index == 0 || index == get_size() - 1) {
+      return QValidator::Invalid;
+    }
+    return m_source->set(index - 1, value);
+  }
+
+  QValidator::State insert(const TableHeaderItem::Model& value,
+      int index) override {
+    return m_source->insert(value, index - 1);
+  }
+
+  QValidator::State move(int source, int destination) override {
+    if(source == 0 || destination == 0 || source == get_size() - 1 ||
+        destination == get_size() - 1) {
+      return QValidator::Invalid;
+    }
+    return m_source->move(source - 1, destination - 1);
+  }
+
+  QValidator::State remove(int index) override {
+    if(index == 0 || index == get_size() - 1) {
+      return QValidator::Invalid;
+    }
+    return m_source->remove(index - 1);
+  }
+
+  connection connect_operation_signal(
+      const OperationSignal::slot_type& slot) const override {
+    return m_transaction.connect_operation_signal(slot);
+  }
+
+  void transact(const std::function<void ()>& transaction) override {
+    m_transaction.transact(transaction);
+  }
+
+  void on_operation(const Operation& operation) {
+    visit(operation,
+      [&] (const AddOperation& operation) {
+        m_transaction.push(AddOperation(operation.m_index + 1,
+          operation.get_value()));
+      },
+      [&] (const MoveOperation& operation) {
+        m_transaction.push(MoveOperation(operation.m_source + 1,
+          operation.m_destination + 1));
+      },
+      [&] (const RemoveOperation& operation) {
+        m_transaction.push(RemoveOperation(operation.m_index + 1,
+          operation.get_value()));
+      },
+      [&] (const UpdateOperation& operation) {
+        m_transaction.push(UpdateOperation(operation.m_index + 1,
+          operation.get_previous(), operation.get_value()));
+      });
+  }
+};
+
+EditableTableView::EditableTableView(
+    std::shared_ptr<TableModel> table, std::shared_ptr<HeaderModel> header,
+    std::shared_ptr<TableFilter> table_filter,
+    std::shared_ptr<CurrentModel> current,
+    std::shared_ptr<SelectionModel> selection, ViewBuilder view_builder,
+    Comparator comparator, QWidget* parent)
+    : TableView(std::make_shared<EditableTableModel>(std::move(table), header),
+        std::make_shared<EditableTableHeaderModel>(header),
+        std::move(table_filter), std::move(current), std::move(selection),
+        std::bind_front(&EditableTableView::make_table_item, this,
+          std::move(view_builder)),
+        std::move(comparator), parent) {
+  auto table_header = static_cast<TableHeader*>(static_cast<Box*>(
```

Comment:
Do not code against implementation details, prefer to code against an API. If you need to modify an API then propose the modification and code against it.